### PR TITLE
Implement the "TokenWrapper" contract

### PR DESCRIPTION
**Pull request by duguorong009**
Summary of changes
Changes introduced in this pull request:
  - Implement the "TokenWrapper" contract
 
Reference issue to close (if applicable)
Closes

### CODE REVIEW

File: contracts/tokenwrapper/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Wrap {} => wrap_native(deps, env, info),
```
Comment (drewstone) [inline contracts/tokenwrapper/src/contract.rs:67]: Why is there no wrap another CW20 token and only native? Is this intentional?

**Review by drewstone — CHANGES_REQUESTED** [review]
Missing comments above functions. Please add documentation.

File: contracts/tokenwrapper/src/contract.rs (hunk 1 above)
Comment (duguorong009) [reply contracts/tokenwrapper/src/contract.rs:67]: https://github.com/webb-tools/protocol-cosmwasm/blob/guorong/implement-tokenwrapper/contracts/tokenwrapper/src/contract.rs#L74
This is where the cw20 wrap is done.
The reason of implementing the cw20 wrapping in the `ReceiveMsg` is that it comes from the characteristic of cosmwasm.
In cosmwasm(cw20), the delegated call is performed with this `Send/Receive` interface.
If you want more detail, you can check https://docs.cosmwasm.com/cw-plus/0.9.0/cw20/spec

File: contracts/tokenwrapper/NOTICE [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Copyright 2022 guorong <dudurong009@gmail.com>
```
Comment (drewstone) [inline contracts/tokenwrapper/NOTICE:1]: This needs to be changed to Webb Tools for every similar file please.

File: contracts/tokenwrapper/src/msg.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct InstantiateMsg {
+    pub count: i32,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ExecuteMsg {
```
Comment (drewstone) [inline contracts/tokenwrapper/src/msg.rs:11]: What are these messages for?

**Review by drewstone — CHANGES_REQUESTED** [review]

File: contracts/tokenwrapper/src/msg.rs (hunk 3 above)
Comment (duguorong009) [reply contracts/tokenwrapper/src/msg.rs]: My bad.
Removed in https://github.com/webb-tools/protocol-cosmwasm/pull/30/commits/6544b7dbefc1a0673d4588b43569cf5ec73ae239

File: contracts/tokenwrapper/NOTICE (hunk 2 above)
Comment (duguorong009) [reply contracts/tokenwrapper/NOTICE]: Removed in https://github.com/webb-tools/protocol-cosmwasm/pull/30/commits/6544b7dbefc1a0673d4588b43569cf5ec73ae239

File: contracts/tokenwrapper/src/state.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema, Default)]
+pub struct Supply {
+    /// issued is how many wrapped tokens this contract has issued
+    pub issued: Uint128,
+}
+
+/// Config
+#[derive(Serialize, Deserialize, Debug, PartialEq, JsonSchema)]
+pub struct Config {
+    pub governer: Addr,
```
Comment (drewstone) [inline contracts/tokenwrapper/src/state.rs:17]: `governor`

File: contracts/tokenwrapper/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+        }),
+    };
+    TOKEN_INFO.save(deps.storage, &data)?;
+
+    // set supply to 0
+    let supply = Supply::default();
+    TOTAL_SUPPLY.save(deps.storage, &supply)?;
+
+    // set config
+    let governer = match msg.governer {
```
Comment (drewstone) [inline contracts/tokenwrapper/src/contract.rs:64]: Fix spelling of `governer` to `governor`

File: contracts/tokenwrapper/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+// Check if the cw20 token address is valid in "TOKENS".
+pub fn is_valid_address(deps: DepsMut, token_address: Addr) -> bool {
+    TOKENS.load(deps.storage, token_address).unwrap_or(false)
+}
+
+// Check if the "wrap_amount" is valid.
+pub fn is_valid_wrap_amount(deps: DepsMut, amount: Uint128) -> bool {
+    let total_supply = TOTAL_SUPPLY.load(deps.storage).unwrap().issued;
+    let config = CONFIG.load(deps.storage).unwrap();
+    amount + total_supply <= config.wrapping_limit
```
Comment (drewstone) [inline contracts/tokenwrapper/src/utils.rs:17]: Are there saturating (safe) methods for addition we should consider using?

**Review by drewstone — CHANGES_REQUESTED** [review]
Spelling is off and should be fixed.

File: contracts/tokenwrapper/src/utils.rs (hunk 6 above)
Comment (duguorong009) [reply contracts/tokenwrapper/src/utils.rs:17]: Yes, there is saturating addition for `Uint128`.
Will check & fix not only this one, but also other operations.

File: contracts/tokenwrapper/src/contract.rs (hunk 5 above)
Comment (duguorong009) [reply contracts/tokenwrapper/src/contract.rs]: Done https://github.com/webb-tools/protocol-cosmwasm/pull/30/commits/76e2d7af747cbd94df5ab288c876b01cb238a178

File: contracts/tokenwrapper/src/state.rs (hunk 4 above)
Comment (duguorong009) [reply contracts/tokenwrapper/src/state.rs]: Done https://github.com/webb-tools/protocol-cosmwasm/pull/30/commits/76e2d7af747cbd94df5ab288c876b01cb238a178

File: contracts/tokenwrapper/src/utils.rs (hunk 6 above)
Comment (duguorong009) [reply contracts/tokenwrapper/src/utils.rs]: Done https://github.com/webb-tools/protocol-cosmwasm/pull/30/commits/1f8f2b4d9a6e24b4c0ee5b9075756055faebe54a

**Pull request merged**